[REVIEW] Added function for mapping a C++ type to it's gdf_dtype enum

jrhemstad opened the pull request:
Fixes https://github.com/rapidsai/cudf/issues/527

This was originally part of https://github.com/rapidsai/cudf/pull/694 but I'm moving it to it's own PR as it's orthogonal.

Also improved type_dispatcher tests to ensure that tests will fail if a new gdf_dtype is added without making the appropriate updates to the type_dispatcher and the type_dispatcher test.

This PR also adds a new utility that emulates a `for_each` over the members of a tuple.

- [x] Added `gdf_dtype_of` function for mapping C++ type to `gdf_dtype` enum
- [x] Added unit tests for `gdf_dtype_of`
- [x] Updated `type_dispatcher` tests to ensure that no new `gdf_dtype` can be added without making the required updates to the `type_dispatcher` and it's associated tests
- [x] Added `device_assert` that performs that same as a `assert` in device code, but will not be compiled out in Release builds
- [x] Added throwing an exception to the error cause in `type_dispatcher`

jrhemstad commented:
I'll wait for your review this time @eyalroz  :smile:

eyalroz commented on cpp/src/utilities/tuple_for_each.hpp:
```diff
@@ -0,0 +1,52 @@
+ * limitations under the License.
+ */
+#ifndef TUPLE_FOR_EACH_H
+#define TUPLE_FOR_EACH_H
```
It's an HPP and the include guard is an H...

eyalroz commented on cpp/src/utilities/tuple_for_each.hpp:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
```
You forget Loius Dionne.

eyalroz commented on cpp/src/utilities/tuple_for_each.hpp:
```diff
@@ -0,0 +1,52 @@
+ * element into the callable.
+ *
+ * This implementation comes from Louis Dionne on StackExchange:
+ * https://codereview.stackexchange.com/a/67394
```
You might want to make sure that the license under which SO code is available can be converted into Apache License 2.

jrhemstad replied:
Looking into this.

eyalroz commented on cpp/src/utilities/tuple_for_each.hpp:
```diff
@@ -0,0 +1,52 @@
+ * @tparam Tuple The type of the tuple
+ * @tparam F The type of the callable
+ * @param tuple The tuple to iterate over
+ * @param f The unary callable
```
Aren't we supposed to mark parameters with [in] and [out]? I'm actually against that, but I vaguely remember this being requested.

eyalroz commented on cpp/src/utilities/tuple_for_each.hpp:
```diff
@@ -0,0 +1,52 @@
+}
+}  // namespace detail
+}  // namespace cudf
+#endif
```
Missing newline.

eyalroz commented on cpp/src/utilities/type_dispatcher.hpp:
```diff
@@ -107,14 +112,83 @@ decltype(auto) type_dispatcher(gdf_dtype dtype,
+  static constexpr gdf_dtype value{GDF_STRING};
+};
+
+}  // namespace cudf
```
I don't think you need such a long specialization... this is a constexpr function and we're in C++14. You should be able to use an appropriate tuple of the relevant types, and convert indices to tuple component element types or back and forth.

Of course, that's just prettying up the code and removing a bit of redundancy; if this works, it works.

By the way - I'm a fan of enums instead of static constexpr's, because they're guaranteed never to actually have storage in memory and only be immediate values, but many people disagree. Still, you need to make it `static constexpr const` I think.

jrhemstad replied:
It would appear there's no performance/optimization difference. The only difference is that `static constexpr` members can't be ODR-used.

https://isocpp.org/blog/2016/05/quick-q-static-constexpr-int-vs-old-fashioned-enum-when-and-why

However, this is no longer true in C++17:

https://en.cppreference.com/w/cpp/language/static

> If a const non-inline (since C++17) static data member or a constexpr static data member (since C++11) is odr-used, a definition at namespace scope is still required, but it cannot have an initializer. **This definition is deprecated for constexpr data members (since C++17).**

I find the `static constexpr` members to be more intuitive, especially for those who've never seen the member `enum` pattern before.

eyalroz replied:
The FAQ is only right up until C++17, at which point the program it lists will [compile just fine](https://godbolt.org/) and _allocate memory space_ for the `static constexpr`.

And the worst thing is, you as the author of the class/struct can't _prevent_ the using code from taking the address-of, or a reference-to, that variable.

So despite the aesthetic point, about which I agree - for now I'm sticking to enums to make sure my code doesn't slip from under my figners.

jrhemstad replied:
I agree that this could be done w/o the explicit list of specializations, but in the interest of time I'm going to leave it as-is. If I have an abundance of free time and am feeling motivated later on, I'll update it.

eyalroz commented on cpp/tests/types/dispatcher_test.cu:
```diff
@@ -13,122 +13,157 @@
+#ifdef NDEBUG
+#undef NDEBUG
+#define REDEFINE
+#endif
```
I thought we agreed to drop this...

eyalroz commented on cpp/tests/types/dispatcher_test.cu:
```diff
@@ -13,122 +13,157 @@
+   * tests will fail.
+   *
+   *---------------------------------------------------------------------------**/
+  std::vector<gdf_dtype> supported_dtypes{
```
Don't use `std::vector` for fixed-size sequences; use `std::array`.

eyalroz commented on cpp/src/utilities/type_dispatcher.hpp:
```diff
@@ -107,14 +112,83 @@ decltype(auto) type_dispatcher(gdf_dtype dtype,
+ * @tparam T The type to map to a `gdf_dtype`
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct type_to_gdf_dtype {
```
Why not a function? 
```
template <typename ColumnElement>
constexpr gdf_dtype gdf_dtype_of() { return GDF_invalid; }
```
we no longer need C++98-style template metaprogramming.

jrhemstad replied:
I'm just following suit with what the STL uses for type-traits. 

Is the best practice guidance to use `constexpr` functions for type-traits now?

eyalroz replied:
Actually, that's a good question to put on StackOverflow. On one hand, remember that most of the standard library was written in the early 1990s where you only had struct-based TMP. And since then we've seen a move towards programming instead of meta-programming as much as possible, not the least of which being:

* `trait_t` instead of `trait::type`
* `trait_v` instead of `trait::value` (C++17)
* `if constexpr` instead of SFINAE+`std::enable_if` (C++17)

But on the other hand, there is the issue of tradition; and the fact that we are _not_ in C++17.

jrhemstad replied:
Updated to a function.

jrhemstad commented:
@eyalroz waiting to hear back about the licensing issue, but otherwise would appreciate a second review and approval if you feel it's good to go.

eyalroz commented:
> @eyalroz waiting to hear back about the licensing issue, but otherwise would appreciate a second review and approval if you feel it's good to go.

You're still undefining `#NDEBUG`, which I thought we agreed you can just drop. Other than that it looks ok. I completely get your choice of keeping the specializations of the dtype-to-type trait for lack of time/desire to do it differently.

jrhemstad commented:
Merging is pending on license approval using the tuple for_each implementation from Stack Exchange.

jrhemstad commented:
@eyalroz @harrism okay, I removed the `NDEBUG` stuff and added a `device_assert` that will not be compiled out in Release builds.

I also went ahead and added throwing an exception in the error case for the `type_dispatcher`. This was to fix https://github.com/rapidsai/cudf/issues/527

harrism commented on CHANGELOG.md:
```diff
@@ -64,6 +65,7 @@
 - PR #1054 Added default ctor to SerialTrieNode to overcome Thrust issue in CentOS7 + CUDA10
 - PR #1024 CSV Reader: Add support for hexadecimal integers in integral-type columns
 
+
```
```suggestion

```

harrism commented on cpp/src/utilities/device_assert.cuh:
```diff
@@ -0,0 +1,42 @@
+ *---------------------------------------------------------------------------**/
+#if defined(__CUDACC_ARCH__) && (defined(__clang__) || defined(__GNUC__))
+#define __ASSERT_STR_HELPER(x) #x
+#define device_assert(e)                                          \
```
I would have thought it should be called "release_assert", since the only difference from "assert" is that it is not compiled out in release builds, not that it is device callable. CUDA already has device `assert()`.

jrhemstad replied:
True, my reasoning is that it's only valid to be called from device code since `__assertfail` is a device intrinsic. 

Is `device_release_assert` too verbose?

eyalroz replied:
@jrhemstad : But it does the same thing a `release_assert()` would do on the host-side. We don't have `device_printf()`, for example. So, yes, too verbose.

I sometimes go with a different verb altogether, like "ensure" or "enforce" - so that people reading the code can't make any assumptions about release vs debug behavior. But `release_assert()` is ok too.

jrhemstad commented:
rerun tests
